### PR TITLE
feat: a lowering pass to re-compose ops into aten.linear

### LINKED ISSUE #2398: ✨[Feature] Writing a lowering pass to re-compose ops into `aten.linear`
The discussion can be found here: https://github.com/pytorch/TensorRT/discussions/2296

### PR DESCRIPTION
**Pull request by zewenli98**
# Description

A lowering pass to re-compose ops into `aten.linear`.

A discussion about why we need this lowering pass can be found here: #2296 

Fixes #2398 

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/lowering/passes/lower_linear.py
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/passes/lower_linear.py:59]: The `linear` operator may not necessarily have the same restrictions as the attention operator did, so it may not be necessary to build the fx graph from scratch. For instance, this whole block could instead be:
```python
    def orig(
        input: torch.Tensor, weight: torch.Tensor, bias: torch.Tensor
    ) -> torch.Tensor:
        W_T = torch.ops.aten.permute.default(weight, [1, 0])
        out = torch.ops.aten.addmm.default(bias, input, W_T)
        return out
```
The above would need verification to validate functionality.

File: tests/py/dynamo/lowering/test_aten_lowering_passes.py
Comment (gs-olive) [inline tests/py/dynamo/lowering/test_aten_lowering_passes.py:285]: Unexpected ops include `addmm` and `permute`

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:121]: This PR also requires a converter for `aten.addmm`, since we are no longer decomposing it, so it would be a decrease in coverage otherwise

Comment (gs-olive) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py:121]: Additionally, you can remove the `addmm` lowering pass tests, since the functionality is no longer present

**Review by gs-olive — CHANGES_REQUESTED** [review]
Added comments detailing suggested improvements. Overall the design looks good!

File: py/torch_tensorrt/dynamo/lowering/passes/lower_linear.py
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/lowering/passes/lower_linear.py:59]: Thanks for the details. I'm thinking why we need these restrictions for `attention` but not for `linear`. Is there anything different in `attention`? Can we just define the `orig` like:
```python
def orig(
        q: torch.Tensor, k: torch.Tensor, v: torch.Tensor
    ) -> torch.Tensor:
        outputs = torch.ops.aten._scaled_dot_product_efficient_attention.default(q, k, v, None, False)
        out = operator.getitem(outputs, 0)
        return out
```

Comment (zewenli98) [reply py/torch_tensorrt/dynamo/lowering/passes/lower_linear.py:59]: I just tried to use the `orig` function above, and it also works for `attention`.

Comment (gs-olive) [reply py/torch_tensorrt/dynamo/lowering/passes/lower_linear.py]: Ok that's great - the above is definitely cleaner and the preferred usage. I think I had overlooked it or was working on a Torch nightly where that wasn't the case. If you could add the change to the attention pass as well for this PR, that would be much appreciated.

**Comment by zewenli98** [comment]
@gs-olive Thanks for your review and constructive comments. I have addressed all the issues above. Please take a look when you get the chance. Thanks a lot!

**Comment by zewenli98** [comment]
@gs-olive Can we merge into the main branch if everything looks good?

**Comment by gs-olive** [comment]
@zewenli98 - this one looks good, but would like to wait until later this week to merge this particular change

**Review by gs-olive — APPROVED** [review]
Looks good to me!

**Pull request merged**
